File: app/src/main/java/com/example/myapplication/fragments/ChatFragment.java
package com.example.circlo.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.example.circlo.R;
import com.example.circlo.adapters.ChatAdapter;
import com.example.circlo.models.ChatItem;
import java.util.ArrayList;
import java.util.List;

public class ChatFragment extends Fragment {

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_chat, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        RecyclerView rvChats = view.findViewById(R.id.rv_chats);
        rvChats.setLayoutManager(new LinearLayoutManager(getContext()));
        rvChats.setAdapter(new ChatAdapter(getDummyChats()));
    }

    private List<ChatItem> getDummyChats() {
        List<ChatItem> list = new ArrayList<>();
        list.add(new ChatItem("Reja Touring", "Rak Buku Minimalis", "Apakah barangnya masih tersedia?", "09.15", 3));
        list.add(new ChatItem("Sarah Amelia", "Kursi Kayu Vintage", "Oke, saya ambil besok ya", "08.30", 0));
        list.add(new ChatItem("Budi Santoso", "TV LED 32 inch", "Makasih kak barangnya!", "Kemarin", 0));
        return list;
    }
}

File: app/src/main/java/com/example/myapplication/fragments/ChatFragment.kt
package com.example.myapplication.fragments

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.R
import com.example.myapplication.adapters.ChatRoomAdapter
import com.example.myapplication.chat.ChatRoomActivity
import com.example.myapplication.models.ChatRoom
import com.example.myapplication.network.SupabaseClientProvider
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.launch

class ChatFragment : Fragment() {

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ) = inflater.inflate(R.layout.fragment_chat, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val rv: RecyclerView = view.findViewById(R.id.rv_chats)
        rv.layoutManager = LinearLayoutManager(context)

        val userId = SupabaseClientProvider.client
            .auth.currentSessionOrNull()?.user?.id ?: return

        lifecycleScope.launch {
            try {
                val rooms = SupabaseClientProvider.client
                    .postgrest["chat_rooms"]
                    .select()
                    .decodeList<ChatRoom>()
                    .filter { it.donorId == userId || it.requesterId == userId }
                    .sortedByDescending { it.lastMessageAt }

                rv.adapter = ChatRoomAdapter(rooms) { room ->
                    val intent = Intent(requireContext(), ChatRoomActivity::class.java)
                    intent.putExtra("room_id", room.id)
                    intent.putExtra("other_username", room.donorId)
                    startActivity(intent)
                }
            } catch (e: Exception) {
                Toast.makeText(context, "Error: ${e.message}", Toast.LENGTH_SHORT).show()
            }
        }
    }
}